torchtune/models/llava_next/image_processor.py

```diff
@@ -0,0 +1,344 @@
+
+        return image_patches
+
+
+class ImageProcessor(nn.Module):
+    def __init__(
+        self,
+        patch_size: int = 336,
+        possible_resolutions: Optional[Tuple[Tuple[int, int]]] = None,
+        max_num_chunks: int = 4,
+        resample: str = "bicubic",
+        do_rescale: bool = True,
+        rescale_factor: Union[int, float] = 1 / 255,
+        do_normalize: bool = True,
+        image_mean: Optional[Union[float, Tuple[float]]] = None,
+        image_std: Optional[Union[float, Tuple[float]]] = None,
+    ) -> None:
+        """
+        Args:
+            patch_size (int): Size of the patches to divide the image into.
+            possible_resolutions (Optional[List[Tuple[int, int]]]): List of possible resolutions as tuples (height, width).
+            max_num_chunks (int): Only used possible_resolutions is NOT given. Maximum number of chunks for processing
+            high-resolution images.
+            resample (str): Resampling method used when resizing images.
+            do_rescale (bool): Flag to determine whether to rescale the image by `rescale_factor`.
+            rescale_factor (Union[int, float]): Scale factor used if rescaling the image.
+            do_normalize (bool): Flag to determine whether to normalize the image.
+            image_mean (Optional[Union[float, List[float]]]): Mean values for normalization.
+            image_std (Optional[Union[float, List[float]]]): Standard deviation values for normalization.
+        """
+        super().__init__()
+
+        logger.info("Initializating ImageProcessor...")
+
+        if image_mean is None:
+            image_mean = [0.48145466, 0.4578275, 0.40821073]
+        if image_std is None:
+            image_std = [0.26862954, 0.26130258, 0.27577711]
+
+        # If possible_resolutions are not given, then calculate possible ones based on max_num_chunks
+        if not possible_resolutions:
+            possible_resolutions = find_supported_resolutions(
+                max_num_chunks=max_num_chunks, patch_size=patch_size
+            )
+        else:
+            possible_resolutions = possible_resolutions
+
+        logger.info(f"possible_resolutions: {possible_resolutions}")
+
+        self.patchfy = GetImagePatches(
+            possible_resolutions=possible_resolutions,
+            patch_size=patch_size,
+            resample=resample,
+        )
+
+        _preprocess = []
+        if do_rescale:
+            _preprocess.append(v2.Normalize(mean=[0] * 3, std=[1 / rescale_factor] * 3))
+
+        if do_normalize:
+            _preprocess.append(v2.Normalize(mean=image_mean, std=image_std))
+
+        self._preprocess = nn.Sequential(*_preprocess)
+
+    def preprocess(
+        self, image: ImageInput
+    ) -> Dict[str, torch.Tensor | Tuple[int, int]]:
+
+        # Make image have dimension [3, H, W]. Input can be grayscale, RGB, channels-first or last.
+        image = F.grayscale_to_rgb_image(F.to_image(image))
```

Comment:
why is this class not just `nn.Sequential` of torchvision transforms? the image patch seems to be a custom transform, the `F` functions here are also transforms, and `preprocess` is just normalize.

I know this is coupled with llava next, but if this will be our primary example for image transforms that users will emulate, then we need to consider how they can look at these, copy, and easily extend for their own use case

Reply:
I was debating myself about it.

The reason I didnt add them to nn.Sequential is because they are not compatible with torch.compile / jit.script, but nn.Sequential is.

Alternatively, I can use v2.compose(*transforms) instead of nn.Sequential(*transforms), but then torch.compile / jit.script has no effect (i guess). Thas the difference between compose and sequential.

I tried compiling it, and didnt see any changes in performance, so maybe that doesnt matter? I am happy to discuss it more / change it.

here is the doc: https://pytorch.org/vision/stable/generated/torchvision.transforms.v2.ToImage.html#torchvision.transforms.v2.ToImage